builtin/credential/aws-iam/path_login.go

```diff
@@ -0,0 +1,376 @@
+	if err != nil {
+		return logical.ErrorResponse("Error getting configuration"), nil
+	}
+
+	if config.HeaderValue != "" {
+		ok, msg := ensureVaultHeaderValue(headers, parsedUrl, config.HeaderValue)
+		if !ok {
+			return logical.ErrorResponse(fmt.Sprintf("Error validating %s header: %s", magicVaultHeader, msg)), nil
+		}
+	}
+
+	endpoint := config.Endpoint
+	if endpoint == "" {
+		endpoint = "https://sts.amazonaws.com"
+	}
+
+	clientArn, err := submitCallerIdentityRequest(method, endpoint, parsedUrl, body, headers)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("Error making upstream request: %s", err)), nil
+	}
+	canonicalArn, principalName, err := parseIamArn(clientArn)
+	if err != nil {
+		return logical.ErrorResponse("Unrecognized IAM principal type"), nil
+	}
+
+	roleName := data.Get("role").(string)
+	if roleName == "" {
+		roleName = principalName
+	}
+
+	roleEntry, err := b.lockedAWSRole(req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		return logical.ErrorResponse(fmt.Sprintf("entry for role %s not found", roleName)), nil
+	}
+
+	if roleEntry.BoundIamPrincipal != canonicalArn {
+		return logical.ErrorResponse(fmt.Sprintf("IAM Principal '%s' does not belong to the role '%s'", clientArn, roleName)), nil
+	}
+
+	resp := &logical.Response{
+		Auth: &logical.Auth{
+			Policies: roleEntry.Policies,
+			Metadata: map[string]string{
+				"client_arn":    clientArn,
+				"canonical_arn": canonicalArn,
+			},
+			InternalData: map[string]interface{}{
+				"role_name": roleName,
+			},
+			DisplayName: principalName,
+			LeaseOptions: logical.LeaseOptions{
+				Renewable: true,
+				TTL:       roleEntry.TTL,
+			},
+		},
+	}
+
+	shortestTTL := b.System().DefaultLeaseTTL()
+	if roleEntry.TTL > time.Duration(0) && roleEntry.TTL < shortestTTL {
+		shortestTTL = roleEntry.TTL
+	}
+
+	maxTTL := b.System().MaxLeaseTTL()
+	if roleEntry.MaxTTL > time.Duration(0) && roleEntry.MaxTTL < maxTTL {
+		maxTTL = roleEntry.MaxTTL
+	}
+
+	if shortestTTL > maxTTL {
+		resp.AddWarning(fmt.Sprintf("Effective TTL of %q exceeded the effective max_ttl of %q; TTL value is capped accordingly", (shortestTTL / time.Second).String(), (maxTTL / time.Second).String()))
+		shortestTTL = maxTTL
+	}
+
+	resp.Auth.TTL = shortestTTL
+
+	return resp, nil
+}
+
+func (b *backend) pathLoginRenew(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	canonicalArn := req.Auth.Metadata["canonical_arn"]
+	if canonicalArn == "" {
+		return nil, fmt.Errorf("unable to retrieve canonical ARN from metadata during renewal")
+	}
+
+	roleName := req.Auth.InternalData["role_name"].(string)
+	if roleName == "" {
+		return nil, fmt.Errorf("Error retrieving role_name during renewal")
+	}
+	roleEntry, err := b.lockedAWSRole(req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		return nil, fmt.Errorf("role entry not found")
+	}
+
+	if roleEntry.BoundIamPrincipal != canonicalArn {
```

Comment:
I think the confusing factor here might be this part of the docs:
```
The GetSessionToken action must be called by using the long-term AWS security credentials of the AWS account or an IAM user. Credentials that are created by IAM users are valid for the duration that you specify, from 900 seconds (15 minutes) up to a maximum of 129600 seconds (36 hours), with a default of 43200 seconds (12 hours); credentials that are created by using account credentials can range from 900 seconds (15 minutes) up to a maximum of 3600 seconds (1 hour), with a default of 1 hour.
```
It's not clear to me what the difference is between account credentials and user credentials. So my thinking was that the user could submit a new signed request with the renewal, but if the original token only lasts an hour then it's an issue anyways.